core/templates/services/playthrough.service.ts

```diff
@@ -17,383 +17,269 @@
-      if (this.isPlaythroughDiscarded()) {
-        return;
-      }
-      PlaythroughService.playthrough.actions.push(
-        this.learnerActionObjectFactory.createNew(
-          AppConstants.ACTION_TYPE_EXPLORATION_QUIT,
-          {
-            state_name: {
-              value: stateName
-            },
-            time_spent_in_state_in_msecs: {
-              value: timeSpentInStateSecs
-            }
-          },
-          ServicesConstants.CURRENT_ACTION_SCHEMA_VERSION
-        ));
+    this.misTracker.recordStateTransition(destStateName);
+    this.cstTracker.recordStateTransition(destStateName);
+
+    const answerSubmitAction = (
+      this.learnerActionObjectFactory.createAnswerSubmitAction({
+        state_name: {value: stateName},
+        dest_state_name: {value: destStateName},
+        interaction_id: {value: interactionId},
+        submitted_answer: {value: answer},
+        feedback: {value: feedback},
+        time_spent_state_in_msecs: {value: 1000 * timeSpentInStateSecs}
+      }));
+    this.playthrough.actions.push(answerSubmitAction);
+  }
+
+  recordExplorationQuitAction(
+      stateName: string, timeSpentInStateSecs: number): void {
+    if (this.playthrough === null ||
+        this.playthrough.getLastAction().actionType === 'ExplorationQuit') {
+      return;
     }
 
-    recordPlaythrough(isExplorationComplete: boolean): void {
-      if (this.isPlaythroughDiscarded()) {
-        return;
-      }
-      if (isExplorationComplete) {
-        // If the exploration is completed, do not check for issues.
-        return;
-      }
-      if (PlaythroughService.playthrough.playthroughId) {
-        // Playthrough ID exists, so issue has already been identified.
-        this.removeOldQuitAction();
-        if (PlaythroughService.playthrough.issueType ===
-          AppConstants.ISSUE_TYPE_EARLY_QUIT) {
-          // If the existing issue is of type early quit, and some other issue
-          // can be identified, update the issue since early quit has lower
-          // priority.
-          this.analyzePlaythrough();
-        }
-        this.storePlaythrough(false);
-      } else {
-        // Playthrough ID doesn't exist.
-        this.analyzePlaythrough();
-        if (PlaythroughService.playthrough.issueType) {
-          // Issue type exists, so an issue is identified after analyzing the
-          // playthrough, and the playthrough is stored.
-          this.storePlaythrough(true);
-        }
-      }
+    this.expDurationInSecs = this.stopwatch.getTimeInSecs();
+    this.eqTracker.recordExplorationQuit(stateName, timeSpentInStateSecs);
+
+    const explorationQuitAction = (
+      this.learnerActionObjectFactory.createExplorationQuitAction({
+        state_name: {value: stateName},
+        time_spent_in_state_in_msecs: {value: 1000 * timeSpentInStateSecs}
+      }));
+    this.playthrough.actions.push(explorationQuitAction);
+
+    this.classifyPlaythrough();
+  }
+
+  storePlaythrough(): void {
```

Comment:
Does this take into account the learnerIsInSamplePopulation value? Should it?

Reply:
OK, maybe mention that as a comment, or (perhaps better) add a gate here with an isPlaythroughRecordingEnabled() check? It may not be strictly necessary but seems like a good safety guard in case other folks decide to use this function in the future.